* Association rule mining : 
used to detect the relation between different associations, i.e., used to uncover hidden relations between variables in the large datasets. for example, lets say that most of the customers who gets paneer also gets paneer butter masala, we can use this information to increase sales.

## APRIORI ALGORITHM : 
one of the most popular algorithm used for association rule mining.

    it uses 3 main factors : support, confidence and lift
    
    support :calculates how popular is item[i] in the given dataset 
    support(item[i]) = number of transactions containing item[i] / total number of transactions

    confidence : calculates how likely item[j] is purchased when item[i] is purchased.
    confidence(item[i] -> item[j]) = number of transcations containing item[i] and item[j] / number of transactions containing item[i] => support(item[i, j]) / support(item[i])

* one drawback of confidence is that it only accounts for propularity of item[i] only but it does not take into account how popular item[j] is. if item[j] is also very popular, then there will be higher chances of containing both item[i] and item[j] in the same transaction, thus inflating the confidence level. this is why we use 'lift'

        lift : calculates how like item[j] is purchased when item[i] is purchased, while controlling how    popular item[j] is
        lift(item[i] -> item[j]) = support(item[i, j]) / [support(item[i]) * support(item[j])]

### method 1 - using apyori library

In [71]:
import numpy as np
import pandas as pd

# pip install apyori
from apyori import apriori

In [72]:
df = pd.read_csv('./data-1.csv', header = None)

In [73]:
df.head(10)

0             1           2                 3   \
0             shrimp       almonds     avocado    vegetables mix   
1            burgers     meatballs        eggs               NaN   
2            chutney           NaN         NaN               NaN   
3             turkey       avocado         NaN               NaN   
4      mineral water          milk  energy bar  whole wheat rice   
5     low fat yogurt           NaN         NaN               NaN   
6  whole wheat pasta  french fries         NaN               NaN   
7               soup   light cream     shallot               NaN   
8  frozen vegetables     spaghetti   green tea               NaN   
9       french fries           NaN         NaN               NaN   

             4                 5     6               7             8   \
0  green grapes  whole weat flour  yams  cottage cheese  energy drink   
1           NaN               NaN   NaN             NaN           NaN   
2           NaN               NaN   NaN             NaN           NaN   
3           NaN               NaN   NaN             NaN           NaN   
4     green tea               NaN   NaN             NaN           NaN   
5           NaN               NaN   NaN             NaN           NaN   
6           NaN               NaN   NaN             NaN           NaN   
7           NaN               NaN   NaN             NaN           NaN   
8           NaN               NaN   NaN             NaN           NaN   
9           NaN               NaN   NaN             NaN           NaN   

             9               10         11     12     13             14  \
0  tomato juice  low fat yogurt  green tea  honey  salad  mineral water   
1           NaN             NaN        NaN    NaN    NaN            NaN   
2           NaN             NaN        NaN    NaN    NaN            NaN   
3           NaN             NaN        NaN    NaN    NaN            NaN   
4           NaN             NaN        NaN    NaN    NaN            NaN   
5           NaN             NaN        NaN    NaN    NaN            NaN   
6           NaN             NaN        NaN    NaN    NaN            NaN   
7           NaN             NaN        NaN    NaN    NaN            NaN   
8           NaN             NaN        NaN    NaN    NaN            NaN   
9           NaN             NaN        NaN    NaN    NaN            NaN   

       15                 16               17       18         19  
0  salmon  antioxydant juice  frozen smoothie  spinach  olive oil  
1     NaN                NaN              NaN      NaN        NaN  
2     NaN                NaN              NaN      NaN        NaN  
3     NaN                NaN              NaN      NaN        NaN  
4     NaN                NaN              NaN      NaN        NaN  
5     NaN                NaN              NaN      NaN        NaN  
6     NaN                NaN              NaN      NaN        NaN  
7     NaN                NaN              NaN      NaN        NaN  
8     NaN                NaN              NaN      NaN        NaN  
9     NaN                NaN              NaN      NaN        NaN

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7501 entries, 0 to 7500
Data columns (total 20 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   0       7501 non-null   object
 1   1       5747 non-null   object
 2   2       4389 non-null   object
 3   3       3345 non-null   object
 4   4       2529 non-null   object
 5   5       1864 non-null   object
 6   6       1369 non-null   object
 7   7       981 non-null    object
 8   8       654 non-null    object
 9   9       395 non-null    object
 10  10      256 non-null    object
 11  11      154 non-null    object
 12  12      87 non-null     object
 13  13      47 non-null     object
 14  14      25 non-null     object
 15  15      8 non-null      object
 16  16      4 non-null      object
 17  17      4 non-null      object
 18  18      3 non-null      object
 19  19      1 non-null      object
dtypes: object(20)
memory usage: 1.1+ MB


In [75]:
# filling 'nan' with 0
df.fillna(0, inplace = True)

In [76]:
df.shape

(7501, 20)

In [77]:
# for 'apriori' algorithm, we need to convert the data into list
# transactions = [['item1', 'item2', 'item3'], ['applle'], ['mango', 'watermelon'], [...], ...]


transactions = []
for i in range(df.shape[0]) : 
    temp_list = [x for x in df.iloc[i, :] if x!=0]
    transactions.append(temp_list)

In [78]:
transactions

[['shrimp',
  'almonds',
  'avocado',
  'vegetables mix',
  'green grapes',
  'whole weat flour',
  'yams',
  'cottage cheese',
  'energy drink',
  'tomato juice',
  'low fat yogurt',
  'green tea',
  'honey',
  'salad',
  'mineral water',
  'salmon',
  'antioxydant juice',
  'frozen smoothie',
  'spinach',
  'olive oil'],
 ['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries'],
 ['eggs', 'pet food'],
 ['cookies'],
 ['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
 ['spaghetti', 'champagne', 'cookies'],
 ['mineral water', 'salmon'],
 ['mineral water'],
 ['shrimp',
  'chocolate',
  'chicken',
  'honey',
  'oil',
  'cooking oil',
  'low fat yogurt'],
 ['turkey', 'eggs'],
 ['turkey',
  'fresh tuna',
  'tomatoes',
  'spagh

In [79]:
rules = apriori(transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2)

## min_support = 0.003 -> means selecting items with min support of 0.3%
## min_confidance = 0.2 -> means min confidance of 20% 
## min_lift = 3  
## min_length = 2 -> means no. of items in the transaction should be 2

In [80]:
rules #generator object

<generator object apriori at 0x00000204641849E0>

In [81]:
result = list(rules) #converting generator object to list
result

[RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'escalope', 'pasta'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'fromage blanc', 'honey'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

In [82]:
df_results = pd.DataFrame(result) #saving it in a dataframe
df_results.head()

items   support  \
0            (chicken, light cream)  0.004533   
1  (escalope, mushroom cream sauce)  0.005733   
2                 (escalope, pasta)  0.005866   
3            (fromage blanc, honey)  0.003333   
4      (herb & pepper, ground beef)  0.015998   

                                  ordered_statistics  
0  [((light cream), (chicken), 0.2905982905982905...  
1  [((mushroom cream sauce), (escalope), 0.300699...  
2  [((pasta), (escalope), 0.3728813559322034, 4.7...  
3  [((fromage blanc), (honey), 0.2450980392156863...  
4  [((herb & pepper), (ground beef), 0.3234501347...

In [83]:
support = df_results.support

In [84]:
df_results['ordered_statistics'] # you can notice that this is a list of values

0     [((light cream), (chicken), 0.2905982905982905...
1     [((mushroom cream sauce), (escalope), 0.300699...
2     [((pasta), (escalope), 0.3728813559322034, 4.7...
3     [((fromage blanc), (honey), 0.2450980392156863...
4     [((herb & pepper), (ground beef), 0.3234501347...
                            ...                        
75    [((olive oil, ground beef), (spaghetti, minera...
76    [((pancakes, ground beef), (spaghetti, mineral...
77    [((tomatoes, ground beef), (spaghetti, mineral...
78    [((spaghetti, mineral water, milk), (olive oil...
79    [((tomatoes, milk), (spaghetti, mineral water)...
Name: ordered_statistics, Length: 80, dtype: object

In [85]:
# df_results.opdered_statisitcs has 4 values : lhs, rhs, confidence, lift

lhs = []
rhs = []
confidence = []
lift = []

for i in range(df_results.shape[0]):
    single_list = df_results['ordered_statistics'][i][0]
    lhs.append(list(single_list[0]))
    rhs.append(list(single_list[1]))
    confidence.append(single_list[2])
    lift.append(single_list[3])

In [86]:
# convert all four list into dataframe for further operation

lhs = pd.DataFrame(lhs)
rhs = pd.DataFrame(rhs)
confidance=pd.DataFrame(confidence,columns=['confidence'])
lift=pd.DataFrame(lift,columns=['lift'])

In [87]:
# concat all list together in a single dataframe
df_final = pd.concat([lhs,rhs,support,confidance,lift], axis=1)
df_final

0              1     2            0              1  \
0            light cream           None  None      chicken           None   
1   mushroom cream sauce           None  None     escalope           None   
2                  pasta           None  None     escalope           None   
3          fromage blanc           None  None        honey           None   
4          herb & pepper           None  None  ground beef           None   
..                   ...            ...   ...          ...            ...   
75             olive oil    ground beef  None    spaghetti  mineral water   
76              pancakes    ground beef  None    spaghetti  mineral water   
77              tomatoes    ground beef  None    spaghetti  mineral water   
78             spaghetti  mineral water  milk    olive oil           None   
79              tomatoes           milk  None    spaghetti  mineral water   

     support  confidence      lift  
0   0.004533    0.290598  4.843951  
1   0.005733    0.300699  3.790833  
2   0.005866    0.372881  4.700812  
3   0.003333    0.245098  5.164271  
4   0.015998    0.323450  3.291994  
..       ...         ...       ...  
75  0.003066    0.216981  3.632981  
76  0.003066    0.211009  3.532991  
77  0.003066    0.261364  4.376091  
78  0.003333    0.211864  3.216994  
79  0.003333    0.238095  3.986501  

[80 rows x 8 columns]

In [88]:
df_final.fillna(' ', inplace = True)
df_final

0              1     2            0              1  \
0            light cream                           chicken                  
1   mushroom cream sauce                          escalope                  
2                  pasta                          escalope                  
3          fromage blanc                             honey                  
4          herb & pepper                       ground beef                  
..                   ...            ...   ...          ...            ...   
75             olive oil    ground beef          spaghetti  mineral water   
76              pancakes    ground beef          spaghetti  mineral water   
77              tomatoes    ground beef          spaghetti  mineral water   
78             spaghetti  mineral water  milk    olive oil                  
79              tomatoes           milk          spaghetti  mineral water   

     support  confidence      lift  
0   0.004533    0.290598  4.843951  
1   0.005733    0.300699  3.790833  
2   0.005866    0.372881  4.700812  
3   0.003333    0.245098  5.164271  
4   0.015998    0.323450  3.291994  
..       ...         ...       ...  
75  0.003066    0.216981  3.632981  
76  0.003066    0.211009  3.532991  
77  0.003066    0.261364  4.376091  
78  0.003333    0.211864  3.216994  
79  0.003333    0.238095  3.986501  

[80 rows x 8 columns]

In [89]:
df_final.columns = ['lhs',1, 2, 'rhs',3,'support','confidance','lift']
df_final.head()

lhs  1  2          rhs  3   support  confidance      lift
0           light cream            chicken     0.004533    0.290598  4.843951
1  mushroom cream sauce           escalope     0.005733    0.300699  3.790833
2                 pasta           escalope     0.005866    0.372881  4.700812
3         fromage blanc              honey     0.003333    0.245098  5.164271
4         herb & pepper        ground beef     0.015998    0.323450  3.291994

In [90]:
# combining 1, 2 into lhs and 3 into rhs
df_final['lhs'] = df_final['lhs'] + str(',') + df_final[1] + str(',') + df_final[2]
df_final['rhs'] = df_final['rhs'] + str(',') + df_final[3]
df_final.drop(columns = [1, 2, 3], inplace = True)
df_final

lhs                      rhs   support  \
0                light cream, ,                 chicken,   0.004533   
1       mushroom cream sauce, ,                escalope,   0.005733   
2                      pasta, ,                escalope,   0.005866   
3              fromage blanc, ,                   honey,   0.003333   
4              herb & pepper, ,             ground beef,   0.015998   
..                           ...                      ...       ...   
75       olive oil,ground beef,   spaghetti,mineral water  0.003066   
76        pancakes,ground beef,   spaghetti,mineral water  0.003066   
77        tomatoes,ground beef,   spaghetti,mineral water  0.003066   
78  spaghetti,mineral water,milk              olive oil,   0.003333   
79               tomatoes,milk,   spaghetti,mineral water  0.003333   

    confidance      lift  
0     0.290598  4.843951  
1     0.300699  3.790833  
2     0.372881  4.700812  
3     0.245098  5.164271  
4     0.323450  3.291994  
..         ...       ...  
75    0.216981  3.632981  
76    0.211009  3.532991  
77    0.261364  4.376091  
78    0.211864  3.216994  
79    0.238095  3.986501  

[80 rows x 5 columns]

In [91]:
df_final = df_final.sort_values(by = 'lift', ascending=False)

In [92]:
# top 10 values 
df_final.head(10)

lhs                    rhs  \
70                   frozen vegetables,soup,      mineral water,milk   
69              olive oil,frozen vegetables,      mineral water,milk   
52          whole wheat pasta,mineral water,             olive oil,    
44                           tomato sauce, ,   spaghetti,ground beef   
3                           fromage blanc, ,                 honey,    
0                             light cream, ,               chicken,    
2                                   pasta, ,              escalope,    
24                 french fries,ground beef,         herb & pepper,    
61  frozen vegetables,mineral water,chocolate               shrimp,    
66            frozen vegetables,ground beef,      mineral water,milk   

     support  confidance      lift  
70  0.003066    0.383333  7.987176  
69  0.003333    0.294118  6.128268  
52  0.003866    0.402778  6.115863  
44  0.003066    0.216981  5.535971  
3   0.003333    0.245098  5.164271  
0   0.004533    0.290598  4.843951  
2   0.005866    0.372881  4.700812  
24  0.003200    0.230769  4.665768  
61  0.003200    0.328767  4.600900  
66  0.003733    0.220472  4.593788

### method 2 - using mlxtend.frequent_patterns library to perform apriori algorithm.


the output of apyori.apriori is not the most user friendly. whereas mlxtend's apriori is much more modern and doesn't require much data processing to properly read the output

In [93]:
from mlxtend.frequent_patterns import apriori as mlx_apy
from mlxtend.frequent_patterns import association_rules

In [94]:
data = pd.read_csv('./data-2.csv')
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [95]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [96]:
# removing space start and end of the string if there is any
data.Description = data.Description.str.strip()

In [97]:
# removing the row if quantity < 0 as it's not possible
data = data[data.Quantity > 0]

In [128]:
# we will only take transactions from france 
data = data[data.Country == 'France']

In [130]:
'''
mlxtend.frequent_patterns.apriori requires data in the below format

    transaction_name    apple banana grapes
transaction  1            0     1      1
             2            1     0      1  
             3            1     0      0
             4            0     1      0          
'''

# converting the data using pivot_table()
basket = pd.pivot_table(data = data, index='InvoiceNo', columns='Description', values = 'Quantity', aggfunc='sum', fill_value=0)

basket.head()

Description  10 COLOUR SPACEBOY PEN  12 COLOURED PARTY BALLOONS  \
InvoiceNo                                                         
536370                            0                           0   
536852                            0                           0   
536974                            0                           0   
537065                            0                           0   
537463                            0                           0   

Description  12 EGG HOUSE PAINTED WOOD  12 MESSAGE CARDS WITH ENVELOPES  \
InvoiceNo                                                                 
536370                               0                                0   
536852                               0                                0   
536974                               0                                0   
537065                               0                                0   
537463                               0                                0   

Description  12 PENCIL SMALL TUBE WOODLAND  \
InvoiceNo                                    
536370                                   0   
536852                                   0   
536974                                   0   
537065                                   0   
537463                                   0   

Description  12 PENCILS SMALL TUBE RED RETROSPOT  12 PENCILS SMALL TUBE SKULL  \
InvoiceNo                                                                       
536370                                         0                            0   
536852                                         0                            0   
536974                                         0                            0   
537065                                         0                            0   
537463                                         0                            0   

Description  12 PENCILS TALL TUBE POSY  12 PENCILS TALL TUBE RED RETROSPOT  \
InvoiceNo                                                                    
536370                               0                                   0   
536852                               0                                   0   
536974                               0                                   0   
537065                               0                                   0   
537463                               0                                   0   

Description  12 PENCILS TALL TUBE WOODLAND  ...  WRAP VINTAGE PETALS  DESIGN  \
InvoiceNo                                   ...                                
536370                                   0  ...                            0   
536852                                   0  ...                            0   
536974                                   0  ...                            0   
537065                                   0  ...                            0   
537463                                   0  ...                            0   

Description  YELLOW COAT RACK PARIS FASHION  YELLOW GIANT GARDEN THERMOMETER  \
InvoiceNo                                                                      
536370                                    0                                0   
536852                                    0                                0   
536974                                    0                                0   
537065                                    0                                0   
537463                                    0                                0   

Description  YELLOW SHARK HELICOPTER  ZINC  STAR T-LIGHT HOLDER  \
InvoiceNo                                                         
536370                             0                          0   
536852                             0                          0   
536974                             0                          0   
537065                             0                          0   
537463                             0      

In [131]:
basket['10 COLOUR SPACEBOY PEN'].value_counts()

10 COLOUR SPACEBOY PEN
0     380
24     10
48      2
Name: count, dtype: int64

In [132]:
def convert_into_binary(x):
    if x > 0:
        return 1
    else:
        return 0

basket_sets = basket.applymap(convert_into_binary)

C:\Users\RAJESH\AppData\Local\Temp\ipykernel_13724\2078531064.py:7: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_sets = basket.applymap(convert_into_binary)


In [133]:
basket_sets['10 COLOUR SPACEBOY PEN'].value_counts()

10 COLOUR SPACEBOY PEN
0    380
1     12
Name: count, dtype: int64

before it has different values but now its just 0s and 1s(0 is the item is not in the invoice else 1)

In [134]:
# remove postage item as it is just a seal which almost all transaction contains. 
basket_sets.drop(columns=['POSTAGE'],inplace=True)

In [135]:
# call apriori function and pass minimum support here we are passing 7%. 
# means 7 times in total number of transaction the item should be present.
frequent_itemsets = mlx_apy(basket_sets, min_support=0.07, use_colnames=True)

In [136]:
frequent_itemsets

support                                           itemsets
0   0.071429                      (4 TRADITIONAL SPINNING TOPS)
1   0.096939                       (ALARM CLOCK BAKELIKE GREEN)
2   0.102041                        (ALARM CLOCK BAKELIKE PINK)
3   0.094388                         (ALARM CLOCK BAKELIKE RED)
4   0.081633                     (BAKING SET 9 PIECE RETROSPOT)
5   0.071429                     (CHILDRENS CUTLERY DOLLY GIRL)
6   0.099490                             (DOLLY GIRL LUNCH BOX)
7   0.096939                          (JUMBO BAG RED RETROSPOT)
8   0.076531                       (JUMBO BAG WOODLAND ANIMALS)
9   0.125000                           (LUNCH BAG APPLE DESIGN)
10  0.084184                      (LUNCH BAG DOLLY GIRL DESIGN)
11  0.153061                          (LUNCH BAG RED RETROSPOT)
12  0.119898                        (LUNCH BAG SPACEBOY DESIGN)
13  0.117347                               (LUNCH BAG WOODLAND)
14  0.142857                 (LUNCH BOX WITH CUTLERY RETROSPOT)
15  0.104592                           (MINI PAINT SET VINTAGE)
16  0.102041                  (PACK OF 72 RETROSPOT CAKE CASES)
17  0.081633                          (PAPER BUNTING RETROSPOT)
18  0.168367                    (PLASTERS IN TIN CIRCUS PARADE)
19  0.137755                         (PLASTERS IN TIN SPACEBOY)
20  0.081633                        (PLASTERS IN TIN STRONGMAN)
21  0.170918                 (PLASTERS IN TIN WOODLAND ANIMALS)
22  0.188776                               (RABBIT NIGHT LIGHT)
23  0.096939                      (RED RETROSPOT CHARLOTTE BAG)
24  0.137755                         (RED RETROSPOT MINI CASES)
25  0.071429                         (RED RETROSPOT PICNIC BAG)
26  0.181122                    (RED TOADSTOOL LED NIGHT LIGHT)
27  0.125000                         (REGENCY CAKESTAND 3 TIER)
28  0.086735                  (RETROSPOT TEA SET CERAMIC 11 PC)
29  0.107143                (ROUND SNACK BOXES SET OF 4 FRUITS)
30  0.158163               (ROUND SNACK BOXES SET OF4 WOODLAND)
31  0.076531                (SET/10 RED POLKADOT PARTY CANDLES)
32  0.132653               (SET/20 RED RETROSPOT PAPER NAPKINS)
33  0.137755                      (SET/6 RED SPOTTY PAPER CUPS)
34  0.127551                    (SET/6 RED SPOTTY PAPER PLATES)
35  0.071429                           (SPACEBOY BIRTHDAY CARD)
36  0.125000                               (SPACEBOY LUNCH BOX)
37  0.122449                (STRAWBERRY LUNCH BOX WITH CUTLERY)
38  0.094388                          (TEA PARTY BIRTHDAY CARD)
39  0.073980                           (WOODLAND CHARLOTTE BAG)
40  0.073980  (ALARM CLOCK BAKELIKE GREEN, ALARM CLOCK BAKEL...
41  0.079082  (ALARM CLOCK BAKELIKE RED, ALARM CLOCK BAKELIK...
42  0.073980  (ALARM CLOCK BAKELIKE RED, ALARM CLOCK BAKELIK...
43  0.071429         (SPACEBOY LUNCH BOX, DOLLY GIRL LUNCH BOX)
44  0.089286  (PLASTERS IN TIN CIRCUS PARADE, PLASTERS IN TI...
45  0.102041  (PLASTERS IN TIN WOODLAND ANIMALS, PLASTERS IN...
46  0.104592  (PLASTERS IN TIN WOODLAND ANIMALS, PLASTERS IN...
47  0.102041  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...
48  0.102041  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...
49  0.122449  (SET/6 RED SPOTTY PAPER PLATES, SET/6 RED SPOT...
50  0.099490  (SET/6 RED SPOTTY PAPER PLATES, SET/20 RED RET...

In [137]:
frequent_itemsets.shape

(51, 2)

Generating frequent itemsets from a list of items - 
--

* itemset : combination of antecedent and the consequent. refer below image
![alt text](itemset.png)

above, we created a list of item_sets that is above the minimum support of 7% which means that each of the itemsets have frequency of more than 7%.

![alt text](1_gZujUCrD9nanRr-ETBFtbg.png)

now , we will apply association rules in the frequent_itemset

Generating all possible rules from the frequent itemsets
--

Once the frequent itemsets are generated, identifying rules out of them is comparatively less taxing. Rules are formed by binary partition of each itemset. If {Bread,Egg,Milk,Butter} is the frequent itemset, candidate rules will look like:

(Egg, Milk, Butter → Bread), (Bread, Milk, Butter → Egg), (Bread, Egg → Milk, Butter), (Egg, Milk → Bread, Butter), (Butter→ Bread, Egg, Milk)

From a list of all possible candidate rules, we aim to identify rules that fall above a minimum threshold level (like min_confidence or min_lift).


In [138]:
rules_mlxtend = association_rules(frequent_itemsets, metric = 'lift', min_threshold = 0.5)
rules_mlxtend.head()

antecedents                   consequents  \
0  (ALARM CLOCK BAKELIKE GREEN)   (ALARM CLOCK BAKELIKE PINK)   
1   (ALARM CLOCK BAKELIKE PINK)  (ALARM CLOCK BAKELIKE GREEN)   
2    (ALARM CLOCK BAKELIKE RED)  (ALARM CLOCK BAKELIKE GREEN)   
3  (ALARM CLOCK BAKELIKE GREEN)    (ALARM CLOCK BAKELIKE RED)   
4    (ALARM CLOCK BAKELIKE RED)   (ALARM CLOCK BAKELIKE PINK)   

   antecedent support  consequent support   support  confidence      lift  \
0            0.096939            0.102041  0.073980    0.763158  7.478947   
1            0.102041            0.096939  0.073980    0.725000  7.478947   
2            0.094388            0.096939  0.079082    0.837838  8.642959   
3            0.096939            0.094388  0.079082    0.815789  8.642959   
4            0.094388            0.102041  0.073980    0.783784  7.681081   

   leverage  conviction  zhangs_metric  
0  0.064088    3.791383       0.959283  
1  0.064088    3.283859       0.964734  
2  0.069932    5.568878       0.976465  
3  0.069932    4.916181       0.979224  
4  0.064348    4.153061       0.960466

In [139]:
rules_mlxtend.shape

(26, 10)

reading above data

In [142]:
rules_mlxtend[ (rules_mlxtend['lift'] >= 4) & (rules_mlxtend['confidence'] >= 0.8) ].head().sort_values(by = 'lift', ascending = False) #these are the most popular item combinations. for example see index 18, if red spotty paper plates is bought then there is very high chance that red spotty paper cups is also bought. seems logical.

antecedents                           consequents  \
2        (ALARM CLOCK BAKELIKE RED)          (ALARM CLOCK BAKELIKE GREEN)   
3      (ALARM CLOCK BAKELIKE GREEN)            (ALARM CLOCK BAKELIKE RED)   
18  (SET/6 RED SPOTTY PAPER PLATES)         (SET/6 RED SPOTTY PAPER CUPS)   
19    (SET/6 RED SPOTTY PAPER CUPS)       (SET/6 RED SPOTTY PAPER PLATES)   
16  (SET/6 RED SPOTTY PAPER PLATES)  (SET/20 RED RETROSPOT PAPER NAPKINS)   

    antecedent support  consequent support   support  confidence      lift  \
2             0.094388            0.096939  0.079082    0.837838  8.642959   
3             0.096939            0.094388  0.079082    0.815789  8.642959   
18            0.127551            0.137755  0.122449    0.960000  6.968889   
19            0.137755            0.127551  0.122449    0.888889  6.968889   
16            0.127551            0.132653  0.102041    0.800000  6.030769   

    leverage  conviction  zhangs_metric  
2   0.069932    5.568878       0.976465  
3   0.069932    4.916181       0.979224  
18  0.104878   21.556122       0.981725  
19  0.104878    7.852041       0.993343  
16  0.085121    4.336735       0.956140

How to read the above data ?
--

> antecedents and consequents  -> The IF component of an association rule is known as the antecedent. The THEN component is known as the consequent. The antecedent and the consequent are disjoint; they have no items in common.

> Example of antecedents and consequents -> Consider antecedents as X and consequents as Y. 

![alt text](itemset.png)

> antecedent support -> This measure gives an idea of how frequent antecedent is in all the transactions. Like (ALARM CLOCK BAKELIKE RED) is present in 9.4% of the transactions. 	

> consequent support -> This measure gives an idea of how frequent consequent is in all the transactions. Like (ALARM CLOCK BAKELIKE GREEN) is present in 9.6% of the transactions. 

> support -> This measure gives an idea of how frequent `ItemSet` is in all the transactions. Like {ALARM CLOCK BAKELIKE RED,ALARM CLOCK BAKELIKE GREEN} is present in 7.9% of the transactions. 

> confidance -> This measure defines the likeliness of occurrence of consequent on the cart given that the cart already has the antecedents. So {ALARM CLOCK BAKELIKE RED} -> {ALARM CLOCK BAKELIKE GREEN} as a confidence of 83%. In simple words their is an 83% chance of finding {ALARM CLOCK BAKELIKE GREEN} , if the cart contains {ALARM CLOCK BAKELIKE RED}.

> lift -> This measure defines the likeliness of occurrence of consequent on the cart given that the cart already has the antecedent, but controlling the popularity of consequent. So lift of {ALARM CLOCK BAKELIKE GREEN} w.r.t {ALARM CLOCK BAKELIKE RED} is 8.64. Which is quite good. Any lift value > 1 implies that the Association rule is worth considering.  [ Reference : https://en.wikipedia.org/wiki/Lift_(data_mining)]

> leverage -> leverage(X -> Y) = P(X and Y) - (P(X)P(Y))

*Leverage measures the difference of X and Y appearing together in the data set and what would be expected if X and Y where statistically dependent. The rational in a sales setting is to find out how many more units (items X and Y together) are sold than expected from the independent sells.*  [Reference : https://michael.hahsler.net/research/recommender/associationrules.html]

> conviction -> Ignore this parameter.  Not much of use in most situations.
